ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol/Abstract.hs

```diff
@@ -0,0 +1,208 @@
+import           Ouroboros.Consensus.Protocol.Abstract (CanBeLeader,
+                     ChainDepState, ConsensusConfig, ConsensusProtocol,
+                     IsLeader, LedgerView, ValidateView)
+import           Ouroboros.Consensus.Protocol.Ledger.HotKey (HotKey)
+import           Ouroboros.Consensus.Protocol.Signed (SignedHeader)
+import           Ouroboros.Consensus.Ticked (Ticked)
+import           Ouroboros.Consensus.Util.Condense (Condense (..))
+
+{-------------------------------------------------------------------------------
+  Crypto
+-------------------------------------------------------------------------------}
+
+type family ProtoCrypto proto :: Type
+
+{-------------------------------------------------------------------------------
+  Header hash
+-------------------------------------------------------------------------------}
+
+newtype ShelleyHash crypto = ShelleyHash
+  { unShelleyHash :: SL.Hash crypto EraIndependentBlockHeader
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (NoThunks)
+
+deriving newtype instance
+  ( Crypto crypto) =>
+  FromCBOR (ShelleyHash crypto)
+
+deriving newtype instance
+  ( Crypto crypto) =>
+  ToCBOR (ShelleyHash crypto)
+
+instance
+  ( Crypto crypto) =>
+  Serialise (ShelleyHash crypto)
+  where
+  encode = toCBOR
+  decode = fromCBOR
+
+
+instance Condense (ShelleyHash crypto) where
+  condense = show . unShelleyHash
+
+{-------------------------------------------------------------------------------
+  Header
+-------------------------------------------------------------------------------}
+
+-- | Shelley header, determined by the associated protocol.
+--
+type family ShelleyProtocolHeader proto = (sh :: Type) | sh -> proto
+
+-- | Indicates that the header (determined by the protocol) supports "Envelope"
+-- functionality. Envelope functionality refers to the minimal functionality
+-- required to construct a chain.
+class
+  ( Eq (EnvelopeCheckError proto),
+    NoThunks (EnvelopeCheckError proto),
+    Show (EnvelopeCheckError proto)
+  ) =>
+  ProtocolHeaderSupportsEnvelope proto
+  where
+  pHeaderHash :: ShelleyProtocolHeader proto -> ShelleyHash (ProtoCrypto proto)
+  pHeaderPrevHash :: ShelleyProtocolHeader proto -> PrevHash (ProtoCrypto proto)
+  pHeaderBodyHash :: ShelleyProtocolHeader proto -> Hash (ProtoCrypto proto) EraIndependentBlockBody
+  pHeaderSlot :: ShelleyProtocolHeader proto -> SlotNo
+  pHeaderBlock :: ShelleyProtocolHeader proto -> BlockNo
+  pHeaderSize :: ShelleyProtocolHeader proto -> Natural
+  pHeaderBlockSize :: ShelleyProtocolHeader proto -> Natural
+
+  type EnvelopeCheckError proto :: Type
+
+  -- | Untick the ledger view.
+  --
+  --   This is needed to implement 'envelopeChecks', hence its inclusion here.
+  --   TODO Find a better place for this
+  untickLedgerView ::
+    proxy proto ->
+    Ticked (LedgerView proto) ->
+    LedgerView proto
+
+  -- | Carry out any additional checks on the envelope
```

Comment:
In what sense are they "additional"? EG beyond the fundamental things like prevhash being correct, slot number being greater, etc?